src/dissectors/ec_o5logon.c

```diff
@@ -62,106 +99,261 @@ FUNC_DECODER(dissector_o5logon)
    void *ident = NULL;
    char tmp[MAX_ASCII_ADDR_LEN];
-   struct o5logon_status *conn_status;
+   struct o5logon_status *conn_status = NULL;
 
    //suppress unused warning
    (void)end;
-   (void) DECODE_DATA;
-   (void) DECODE_DATALEN;
-   (void) DECODED_LEN;
+   (void)DECODE_DATA;
+   (void)DECODE_DATALEN;
+   (void)DECODED_LEN;
+
+   dissect_create_ident(&ident, PACKET, DISSECT_CODE(dissector_o5logon));
 
    if (FROM_CLIENT("o5logon", PACKET)) {
 
-      /* Interesting packets have len >= 4 */
-      if (PACKET->DATA.len < 13)
+      if (session_get(&s, ident, DISSECT_IDENT_LEN) == -E_NOTFOUND) {
+         dissect_create_session(&s, PACKET, DISSECT_CODE(dissector_o5logon));
+         SAFE_CALLOC(s->data, 1, sizeof(struct o5logon_status));
+         session_put(s);
+      }
+
+      /* Fetch state of existing session */
+      conn_status = (struct o5logon_status*)s->data;
+
+      if (PACKET->DATA.len == 0) {
+         /*
+          * We output this packet just in case, but we may see more data
+          * later and output another copy with more fields. See issue #741.
+          */
+         if (conn_status->flags.pkcs != NO && conn_status->flags.user && conn_status->flags.s_sk && conn_status->flags.vfr) {
+            DISSECT_MSG("%s@%s:$o5logon$%s*%s\n", conn_status->user, conn_status->srv_addr, conn_status->srv_sk, conn_status->salt);
+         }
+         if (PACKET->L4.flags & (TH_FIN | TH_RST)) {
+            dissect_wipe_session(PACKET, DISSECT_CODE(dissector_o5logon));
+         }
+         SAFE_FREE(ident);
          return NULL;
+      }
 
-      dissect_create_ident(&ident, PACKET, DISSECT_CODE(dissector_o5logon));
-      /* if the session does not exist... */
-      if (session_get(&s, ident, DISSECT_IDENT_LEN) == -E_NOTFOUND) {
-         unsigned char *skp  = memmem(ptr, PACKET->DATA.len, "AUTH_SESSKEY", 12);
-         unsigned char *sp = memmem(ptr, PACKET->DATA.len, "AUTH_TERMINAL", 13);
-         if (sp && !skp) {
-            /* create the new session */
-            dissect_create_session(&s, PACKET, DISSECT_CODE(dissector_o5logon));
-
-            /* remember the state (used later) */
-            SAFE_CALLOC(s->data, 1, sizeof(struct o5logon_status));
-
-            conn_status = (struct o5logon_status *) s->data;
-            conn_status->status = WAIT_RESPONSE;
-
-            /* find username */
-            unsigned char *last = sp - 6;
-            while(last > ptr) {
-              if(*last == 0xff || *last == 0x01) {
-                 break;
-              }
-              last--;
-            }
-            unsigned int length = *(last+1);
-            if (length < sizeof(conn_status->user))
-            {
-               strncpy((char*)conn_status->user, (char*)last + 2, length);
-               conn_status->user[length] = 0;
-            }
+      /* Interesting packets have len >= 24 */
+      if (PACKET->DATA.len < 24) {
+         SAFE_FREE(ident);
+         return NULL;
+      }
+
+      u_char *ano;
```

Comment:
Please declare variables only at the beginning of a block, otherwise this will generate a warning.